js-pkg/client/src/indexeddb.ts

```diff
@@ -0,0 +1,206 @@
+        Y.applyUpdate(this.doc, update.value)
+
+        this.lastUpdateKey = update.key
+      }
+    }, this)
+  }
+
+  destroy() {
+    this.doc.off('update', this.handleUpdate)
+    this.broadcastChannel.close()
+  }
+
+  async handleUpdate(update: Uint8Array, origin: any) {
+    if (origin === this) {
+      return
+    }
+
+    // We attempt to write in a loop. If we are preempted by another writer, we load the latest
+    // updates and try again.
+    while (true) {
+      await this.loadFromDb()
+      let key = this.updateKey()
+      let newCount = await this.insertValue(key, update)
+
+      if (newCount === null) {
+        // Another writer wrote before we could; reload and try again.
+        continue
+      }
+
+      if (newCount > MAX_UPDATES_IN_STORE) {
+        key = this.updateKey()
+        if (!(await this.saveWholeState(key))) {
+          // Another writer wrote before we could; reload and try again.
+          continue
+        }
+      }
+
+      break
+    }
+
+    this.broadcastChannel.postMessage(this.lastUpdateKey)
+  }
+
+  async getAllValues(range?: IDBKeyRange): Promise<Array<BytesWithKey>> {
+    let transaction = this.db.transaction(OBJECT_STORE_NAME)
+    let objectStore = transaction.objectStore(OBJECT_STORE_NAME)
+    const request = objectStore.getAll(range)
+
+    let result = await new Promise<Array<BytesWithKey>>((resolve, reject) => {
+      request.onsuccess = () => {
+        resolve(request.result)
+      }
+      request.onerror = reject
+    })
+
+    return await Promise.all(
+      result.map(async (data) => {
+        let value = await decryptData(data.value, this.encryptionKey)
+
+        return {
+          key: data.key,
+          value,
+        }
+      }),
+    )
+  }
+
+  async saveWholeState(key: number): Promise<boolean> {
+    const update = Y.encodeStateAsUpdate(this.doc)
+    const encryptedUpdate = await encryptData(update, this.encryptionKey)
+    let transaction = this.db.transaction(OBJECT_STORE_NAME, 'readwrite')
+    let objectStore = transaction.objectStore(OBJECT_STORE_NAME)
+
+    if (await this.hasValue(objectStore, key)) {
+      return false
+    }
+
+    let range = IDBKeyRange.upperBound(key, false)
+    objectStore.delete(range)
+
+    objectStore.add({
+      key,
+      value: encryptedUpdate,
+    })
+
+    return true
+  }
+
+  async hasValue(objectStore: IDBObjectStore, key: number): Promise<boolean> {
+    const request = objectStore.get(key)
+    await new Promise<void>((resolve, reject) => {
+      request.onsuccess = () => resolve()
+      request.onerror = reject
+    })
+
+    return request.result !== undefined
+  }
+
+  /**
+   * Insert a value into IndexedDB. Return the new count of updates in the store if the value was inserted,
+   * or null if the desired key already exists.
+   **/
+  async insertValue(key: number, value: Uint8Array): Promise<number | null> {
+    const encryptedValue = await encryptData(value, this.encryptionKey)
```

Comment:
How expensive is this step? Is it worth putting after the `hasValue` check since its output may not be used?

Reply:
Unfortunately, that's not possible -- WebCrypto APIs are async, and an IndexedDB transaction will be closed as soon as you `await` over a non-IndexedDB promise. We need `hasValue` to happen in the same transaction as `objectStore.put` for correctness reasons, so we are stuck doing this.

Reply:
oh gotcha - didn't realize all these lines could be part of the same transaction!